### PR TITLE
gdsfactory-gen generators and DRC parsing

### PR DESCRIPTION
**Pull request by alibillalhammoud**
This PR provides an intermediate step towards the goals described in [issue210](https://github.com/idea-fasoc/OpenFASOC/issues/210). This introduces the following generators and features:

- mimcap gen
- fet: nmos, pmos gen
- routes: straight, C, L routes
- guardring gen
- via gen
- opamp gen
- pad wrapper for adding pads to opamp (tapeout)
- DRC parser from @ryanrocket

### CODE REVIEW

**Comment by msaligane** [comment]
@joamatab Do you think you can help me review this?

**Comment by joamatab** [comment]
gdsfactory Pcells should follow this pattern to avoid uwanted  changes
we lock changes for a reason :) and you are not supposed to unlock them and modify them after they are in the cache :)

https://gdsfactory.github.io/gdsfactory/notebooks/workflow_0_layout_summary.html?highlight=summary

This is the general pattern we follow
```
@gf.cell
def mzi_with_bend(radius: float = 10):
    """Returns MZI interferometer with bend."""
    c = gf.Component() # the top level component (our canvas)
    mzi = c.add_ref(gf.components.mzi()) # first reference
    bend = c.add_ref(gf.components.bend_euler(radius=radius)) # second reference
    bend.connect("o1", mzi.ports["o2"]) # connect

    c.add_port("o1", port=mzi.ports["o1"]) # expose some ports
    c.add_port("o2", port=bend.ports["o2"]) # expose some ports
    return c
```

overall i'd recommend adding the skywater130 and gf180 pdks into the gdsfactory generic PDK
that's the way we can use the generic PCells into more than 10 PDKs that we currently support from gdsfactory

**Review by saicharan0112 — COMMENTED** [review]
@alibillalhammoud @vijayshankarr I see a lot of development happening to this generator but don't see anything at the documentation level. Not sure whether this is too early to write down something for gdsfactory-gen but would be good if something is started, for eg, explaining the flow or something?

**Comment by saicharan0112** [comment]
@alibillalhammoud can you please mark an order for your PRs?

**Comment by msaligane** [comment]
@alibillalhammoud have you synced your Fork? There are too many changes to review. Let's go over this today.

**Comment by alibillalhammoud** [comment]
@saicharan0112 I just have this PR to merge.

@msaligane I have synced my fork with main. We can go over today.

File: openfasoc/generators/gdsfactory-gen/pygen/README.md
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/pygen/README.md:1]: @alibillalhammoud can we rename pygen as discussed and add this to the docs
cc @saicharan0112

File: openfasoc/generators/gdsfactory-gen/deprecated/scripts/mimcap-array/sky130_fd_pr__cap_mim_m3_2_FJK8MM.gds
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/deprecated/scripts/mimcap-array/sky130_fd_pr__cap_mim_m3_2_FJK8MM.gds:1]: @alibillalhammoud are these gds files using somewhere else? it looks like you moved them in deprecated.

File: openfasoc/generators/gdsfactory-gen/pygen/common/two_transistor_interdigitized.py
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/pygen/common/two_transistor_interdigitized.py:1]: Where is this documented?

File: openfasoc/generators/gdsfactory-gen/pygen/docs/straight_route_def_beh.png
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/pygen/docs/straight_route_def_beh.png:1]: Can we have a better diagram?

File: openfasoc/generators/gdsfactory-gen/pygen/docs/tapring_v1_tree.txt
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/pygen/docs/tapring_v1_tree.txt:1]: Would be good to add a couple comments regarding these port trees

File: openfasoc/generators/gdsfactory-gen/pygen/pdk/gf180_mapped/gf180mcu.drc
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/pygen/pdk/gf180_mapped/gf180mcu.drc:1]: Did we create a github issue to track this. This file should be copied from the pdk folder

File: openfasoc/generators/gdsfactory-gen/pygen/pdk/sky130_mapped/sky130_add_npc.py
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/pygen/pdk/sky130_mapped/sky130_add_npc.py:1]: can we add a description of this fct.

File: openfasoc/generators/gdsfactory-gen/pygen/pdk/util/port_utils.py
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/pygen/pdk/util/port_utils.py:1]: Seems like there are no descriptions to your fcts. So it is hard to follow.

File: openfasoc/generators/gdsfactory-gen/tapeout_and_RL/sky130A/sky130A.magicrc
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/tapeout_and_RL/sky130A/sky130A.magicrc:1]: This should come from the PDK

File: openfasoc/generators/gdsfactory-gen/tapeout_and_RL/sky130_mpw5_pad.gds
Comment (msaligane) [inline openfasoc/generators/gdsfactory-gen/tapeout_and_RL/sky130_mpw5_pad.gds:1]: can you create a folder for gds files

**Review by msaligane — CHANGES_REQUESTED** [review]
Can you please address my comments.

**Comment by msaligane** [comment]
@harshkhandeparkar
FYI

File: openfasoc/generators/gdsfactory-gen/deprecated/scripts/mimcap-array/sky130_fd_pr__cap_mim_m3_2_FJK8MM.gds
Comment (alibillalhammoud) [reply openfasoc/generators/gdsfactory-gen/deprecated/scripts/mimcap-array/sky130_fd_pr__cap_mim_m3_2_FJK8MM.gds:1]: Not sure what those are for, I did not modify the "deprecated" directories at all. If we want we can delete the deprecated directories. They will be in the git history anyways

File: openfasoc/generators/gdsfactory-gen/pygen/pdk/gf180_mapped/gf180mcu.drc
Comment (alibillalhammoud) [reply openfasoc/generators/gdsfactory-gen/pygen/pdk/gf180_mapped/gf180mcu.drc:1]: @msaligane see [issue 235](https://github.com/idea-fasoc/OpenFASOC/issues/235). This tracks all the changes I need to make.

**Comment by saicharan0112** [comment]
@alibillalhammoud I am not sure if this is part of the plan (I don't see it in the tasks) but try to create a Jupyter notebook on how to work with pygen-openfasoc. I think that's the best documentation someone could start with

File: openfasoc/generators/gdsfactory-gen/pygen/pdk/mappedpdk.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        """Construct a mapped pdk from an existing pdk and the extra parts of MappedPDK
+        grid is the grid size in nm"""
+        # input type and value validation
+        if not isinstance(gfpdk, Pdk):
+            raise TypeError("from_gf_pdk: gfpdk arg only accepts GDSFactory pdk type")
+        # create argument dictionary
+        passargs = dict()
+        # pdk args
+        passargs["name"]=gfpdk.name
+        #passargs["cross_sections"]=gfpdk.cross_sections
```
Comment (harshkhandeparkar) [inline openfasoc/generators/gdsfactory-gen/pygen/pdk/mappedpdk.py:239]: Why are these lines commented?

File: openfasoc/generators/gdsfactory-gen/pygen/pdk/sky130_mapped/grules.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
+from ..mappedpdk import MappedPDK
+
+grulesobj = dict()
+for glayer in MappedPDK.valid_glayers:
+    grulesobj[glayer] = dict((x, None) for x in MappedPDK.valid_glayers)
+
+grulesobj["dnwell"]["dnwell"] = {"min_width": 3.0, "min_separation": 6.3}
```
Comment (harshkhandeparkar) [inline openfasoc/generators/gdsfactory-gen/pygen/pdk/sky130_mapped/grules.py:7]: Are these values manually entered, or are they pulled from some file? If they are pulled from a file from the PDK, can this process also be automated?

File: openfasoc/generators/gdsfactory-gen/pygen/pdk/sky130_mapped/sky130_add_npc.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
+from gdsfactory.component import Component
+from gdsfactory.polygon import Polygon
+from gdsfactory.geometry.boolean import boolean
+
+
+# slightly slower version
+#def sky130_add_npc(comp: Component) -> Component:
```
Comment (harshkhandeparkar) [inline openfasoc/generators/gdsfactory-gen/pygen/pdk/sky130_mapped/sky130_add_npc.py:7]: There are some lines of commented code here. Is this code required? If not, it would be best to remove these lines to prevent confusion. If this is an alternative implementation, perhaps they can be combined and an argument can be used to switch between them?

File: openfasoc/generators/gdsfactory-gen/pygen/pdk/util/print_rules.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
+"""read rule deck we have saved in google sheets into a better format for writing to python:
+this is the google sheets format
+https://docs.google.com/spreadsheets/d/172P-CW_dGQU5icyUQAasLVq7Jlz51R5lB1rJ4JyvfIc/edit?usp=sharing
```
Comment (harshkhandeparkar) [inline openfasoc/generators/gdsfactory-gen/pygen/pdk/util/print_rules.py:3]: Seems like this file is inaccessible.

File: openfasoc/generators/gdsfactory-gen/pygen/routing/c_route.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+	e2_extension_comp.add_ports(e2_extension.get_ports_list())
+	me1 = e1_extension_comp << viastack1
+	me2 = e2_extension_comp << viastack2
+	route_ports = [None,None]
+	via_flush = abs((width1 - evaluate_bbox(viastack1)[0])/2) if viaoffset else 0
+	via_flush1 = via_flush if viaoffset[0] else 0-via_flush
+	via_flush1 = 0 if viaoffset[0] is None else via_flush1
+	via_flush2 = via_flush if viaoffset[1] else 0-via_flush
+	via_flush2 = 0 if viaoffset[1] is None else via_flush2
+	if round(edge1.orientation) == 0:# facing east
```
Comment (harshkhandeparkar) [inline openfasoc/generators/gdsfactory-gen/pygen/routing/c_route.py:151]: It seems like the code under this function is repeated many times. Is there a possibility for refactoring?

File: openfasoc/generators/gdsfactory-gen/tapeout_and_RL/sky130_nist_tapeout.py
Comment (harshkhandeparkar) [inline openfasoc/generators/gdsfactory-gen/tapeout_and_RL/sky130_nist_tapeout.py:1]: This is a 1000-line file. I would suggest splitting it for readability.

File: openfasoc/generators/ldo-gen/flow/scripts/openfasoc/custom_place.tcl
[the remark this answers is not in the document]
Comment (harshkhandeparkar) [reply openfasoc/generators/ldo-gen/flow/scripts/openfasoc/custom_place.tcl:1]: If the files get changed each time, you can use `git add gdsfactory-gen` instead of `git add .`.

File: openfasoc/generators/gdsfactory-gen/requirements.txt
[the remark this answers is not in the document]
Comment (alibillalhammoud) [reply openfasoc/generators/gdsfactory-gen/requirements.txt:1]: @harshkhandeparkar did you make a gdsfactory issue about the broken pip dependencies for the sky130 and gf180 packages?

File: openfasoc/generators/gdsfactory-gen/pygen/pdk/sky130_mapped/sky130_add_npc.py (hunk 3 above)
Comment (alibillalhammoud) [reply openfasoc/generators/gdsfactory-gen/pygen/pdk/sky130_mapped/sky130_add_npc.py:7]: yes we can remove

File: openfasoc/generators/gdsfactory-gen/pygen/pdk/util/print_rules.py (hunk 4 above)
Comment (alibillalhammoud) [reply openfasoc/generators/gdsfactory-gen/pygen/pdk/util/print_rules.py:3]: fixed?

Comment (harshkhandeparkar) [reply openfasoc/generators/gdsfactory-gen/pygen/pdk/util/print_rules.py:3]: Yes it is accessible now.

File: openfasoc/generators/gdsfactory-gen/requirements.txt
[the remark this answers is not in the document]
Comment (harshkhandeparkar) [reply openfasoc/generators/gdsfactory-gen/requirements.txt:1]: I haven't yet, but I will soon. I want to test more to figure out what the exact issue is and make sure it is not a problem with my system.

**Comment by saicharan0112** [comment]
> @alibillalhammoud I am not sure if this is part of the plan (I don't see it in the tasks) but try to create a Jupyter notebook on how to work with pygen-openfasoc. I think that's the best documentation someone could start with

I think I just saw one of the issues on documentation for gdsfactory-gen that includes python notebook.

**Review by msaligane — APPROVED** [review]
Approved. But we need to address issues. #235

File: openfasoc/generators/gdsfactory-gen/requirements.txt
[the remark this answers is not in the document]
Comment (saicharan0112) [reply openfasoc/generators/gdsfactory-gen/requirements.txt:1]: I was looking for this exact comment. Is still the old gdsfactory used?. I see that this commit has something about the gdsfactory version in its message.

[the remark this answers is not in the document]
Comment (saicharan0112) [reply openfasoc/generators/gdsfactory-gen/requirements.txt:1]: Also where exactly the packages mentioned inside this `requirements.txt` file is installed?

[the remark this answers is not in the document]
Comment (harshkhandeparkar) [reply openfasoc/generators/gdsfactory-gen/requirements.txt:1]: This `requirements.txt` file just contains dependencies necessary for the gdsfactory generators. They are not installed separately, just listed in a separate file. But for some reason, the root `requirements.txt` also has `gdsfactory` mentioned as a dependency which it should probably not be.
